src/spyglass/spikesorting/v0/spikesorting_burst.py

```diff
@@ -0,0 +1,393 @@
+            timestamp_ind = np.argsort(timestamp)
+            peak_amps[unit_id] = self._get_peak_amps1(
+                waves, unit_id, timestamp_ind
+            )
+            peak_timestamps[unit_id] = timestamp[timestamp_ind]
+
+        self._peak_amp_cache[key_hash] = peak_amps, peak_timestamps
+
+        return peak_amps, peak_timestamps
+
+    def _compute_correlograms(
+        self, key: dict, params: dict = None
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """Compute cross-correlograms for a given key, caching the result.
+
+        Parameters
+        ----------
+        key : dict
+            key of BurstPair
+        params : dict, optional
+            parameters for the computation, default None will check params table
+
+        Returns
+        -------
+        ccgs : np.ndarray
+            cross-correlograms
+        bins : np.ndarray
+            bin edges for the correlograms
+        """
+        key_hash = dj.hash.key_hash(key)
+        if cached := self._xcorrel_cache.get(key_hash):
+            return cached
+        if not params:
+            params = BurstPairParams().get_params(key)
+
+        ccgs, bins = compute_correlograms(
+            waveform_or_sorting_extractor=Curation.get_curated_sorting(key),
+            load_if_exists=False,
+            window_ms=params.get("correl_window_ms", 100.0),
+            bin_ms=params.get("correl_bin_ms", 5.0),
+            method=params.get("correl_method", "numba"),
+        )
+
+        self._xcorrel_cache[key_hash] = ccgs, bins
+
+        return ccgs, bins
+
+    def make(self, key) -> None:
+        """Generate BurstPair metrics for a given key"""
+        params = BurstPairParams().get_params(key)
+
+        peak_amps, peak_timestamps = self.get_peak_amps(key)
+        units = peak_amps.keys()
+        if len(units) == 0:
+            self._null_insert(key)
+            return
+
+        # mean waveforms in a dict: each one is of spike number x 4
+        waves = self._get_waves(key)
+        waves_mean_1d = {
+            u: np.reshape(
+                np.mean(waves.get_waveforms(u), axis=0).T,
+                (1, -1),
+            ).ravel()
+            for u in units
+        }
+
+        # calculate cross-correlogram and asymmetry
+        ccgs, bins = self._compute_correlograms(key, params)
+
+        unit_pairs = []
+        for u1, u2 in permutations(units, 2):
+            unit_pairs.append(
+                {
+                    **key,
+                    "unit1": u1,
+                    "unit2": u2,
+                    "wf_similarity": stats.pearsonr(
+                        waves_mean_1d[u1], waves_mean_1d[u2]
+                    ).statistic,
+                    "isi_violation": calculate_isi_violation(
+                        peak_timestamps[u1], peak_timestamps[u2]
+                    ),
+                    "xcorrel_asymm": calculate_ca(
+                        bins[1:], ccgs[u1 - 1, u2 - 1, :]
```

Comment:
Does this assume unit numbers are the same as the index?

Reply:
Changed `id` -> `idx`/`indices` as appropriate